packages/aws-cdk-lib/aws-route53/lib/record-set.ts

```diff
@@ -406,8 +406,15 @@ export class RecordSet extends Resource implements IRecordSet {
     }
 
     if (this.weight !== undefined) {
-      const idPrefix = `WEIGHT_${this.weight}_ID_`;
-      return this.createIdentifier(idPrefix);
+      if (Token.isUnresolved(this.weight)) {
+        const replacement = 'XXX'; // XXX simply because 255 is the highest value for a record weight
+        const idPrefix = `WEIGHT_${replacement}_ID_`;
+        const idTemplate = this.createIdentifier(idPrefix);
+        return idTemplate.replace(replacement, Token.asString(this.weight));
+      } else {
+        const idPrefix = `WEIGHT_${this.weight}_ID_`;
+        return this.createIdentifier(idPrefix);
+      }
```

Comment:
Can you help me understand why this change is needed? I'm a bit worried on this change since it mutate the values when users update CDK version. CloudFormation marks `setIdentifier` property as non-interrupting property which means it won't cause resource replacement, but I'd like to understand this change more to make sure it's a safe change.

Reply:
Hi @GavinZZ sure, I left a comment above trying to explain this. So as a matter of fact, this change is to make sure each synthesis yields the same result with no change.

By nature of supporting a CfnParameter value as the record weight property, the set ID gets kind of broken. This is only the case if a set ID is not provided in the constructor props, so the RecordSet constructor generates its own set ID.

So in the above code, before my change, if a `CfnParameter.valueAsNumber` was provided for the weight property, the actual value would look something like this `-1.8881545897089744e+289`. This would be stringified in the set ID, which significantly reduces the allowed length of the set ID (because we know that a record weight should be a number of 1, 2, or 3 characters in length usually). During synthesis though, this value would get substituted out, but the real problem is the fact that this value is non-deterministic. So on one run the token value for the parameter could be that which I pasted above, but it could be different on a subsequent run. This is what I found when rerunning the integration tests in my local environment. I found sometimes its length could vary by 1 or two characters, and break my tests, since that would affect the generated set ID.

To better understand what I'm trying my best to explain, you might have more luck drilling down into what `RecordSet.createIdentifier` is doing (and the method it calls).

So basically to fix this, because it is not possible to know the parameter value during synthesis time, I am imposing the constraint that if a weight is provided as a CfnParameter, or in other words as an unresolved Token, then we synthetically generate the set ID as if the weight value is 3 characters in length (since that is the maximum length by contract, ie. 255). Then after getting the generated set ID, I replace the dummy part with the token reference (this time using `.valueAsString` instead for better readability), and then we will see it resolved properly post-synthesis and we can be sure that the set ID is reproducible and will not change between seperate syntheses.

I hope that explanation makes sense. Let me know if this is still unclear or if you believe this is actually a breaking change.

Reply:
Thanks for this detailed explanation. I actually totally missed your previous comment. Apologize for that.

I do see that you wrapped the code in an if statement `if (Token.isUnresolved(this.weight))` so it shouldn't affect other use case. My concern is that the final `setIdentifier` value will replace `XXX` with the value of `Token.asString(this.weight)`. 

You're right that the value of `this.weight` (when it's a token) is non-deterministic, and my understanding is that the value `Token.asString(this.weight)` would be non-deterministic either, no?

Reply:
This solution makes sense to me!